src/v/pandaproxy/schema_registry/validation.cc

```diff
@@ -0,0 +1,437 @@
+        if (!has_id) {
+            vlog(
+              plog.debug,
+              "validating: sub: {}, id: {}, has_id: {}",
+              sub,
+              id,
+              has_id);
+            co_return false;
+        }
+
+        _api->_schema_id_cache.local().put(
+          topic, field, sns, id, std::move(proto_offsets));
+        co_return true;
+    };
+
+    ss::future<bool> validate(const model::record_batch& batch) {
+        if (
+          !_record_key_schema_id_validation
+          && !_record_value_schema_id_validation) {
+            co_return true;
+        }
+
+        auto key_is_valid =
+          [this](const model::record& record) -> ss::future<bool> {
+            if (!_record_key_schema_id_validation) {
+                return ss::make_ready_future<bool>(true);
+            }
+            return validate_field(
+              field::key,
+              _topic,
+              _record_key_subject_name_strategy,
+              record.key().copy());
+        };
+
+        auto value_is_valid =
+          [this](const model::record& record) -> ss::future<bool> {
+            if (!_record_value_schema_id_validation) {
+                return ss::make_ready_future<bool>(true);
+            }
+            return validate_field(
+              field::val,
+              _topic,
+              _record_value_subject_name_strategy,
+              record.value().copy());
+        };
+
+        const model::record_batch& b = batch;
+        std::optional<const model::record_batch> u;
+        bool compressed = batch.compressed();
+        if (compressed) {
+            u.emplace(
+              co_await storage::internal::decompress_batch(batch.copy()));
+        }
+
+        bool valid{true};
+        co_await model::for_each_record(
+          compressed ? u.value() : b,
+          [&valid, k{std::move(key_is_valid)}, v{std::move(value_is_valid)}](
+            const model::record& r) {
+              return k(r)
+                .then([&r, v{std::move(v)}](bool k) {
+                    return k ? v(r) : ss::make_ready_future<bool>(false);
+                })
+                .then([&valid](bool v) { valid = v; });
+          });
+        co_return valid;
+    }
+
+    ss::future<bool> validate(const data_t& data) {
+        for (const auto& b : data) {
+            if (!co_await validate(b)) {
+                co_return false;
+            }
+        }
+        co_return true;
+    }
+
+    auto validate(const foreign_data_t& data) { return validate(*data.buffer); }
+
+    ss::future<result> operator()(model::record_batch_reader&& rbr) {
+        if (!_api->_controller->get_feature_table().local().is_active(
+              features::feature::schema_id_validation)) {
+            co_return std::move(rbr);
+        }
+        if (!_api) {
+            // If Schema Registry is not enabled, the safe default is to reject
+            co_return kafka::error_code::invalid_record;
+        }
```

Comment:
This is an issue: Swap the order of these conditionals (check for `!_api` first)

Reply:
I believe the order was intentional. On reconsideration, it's probably safer to swap it.